### ISSUE TITLE
Refactor SpotId

### ISSUE THREAD

**Issue by Zannick** [Zannick/logic-graph#133]
This was a suggestion on the rust users forum. Basically, the idea is that we have a few very large tables that might be impacting compile performance (and potentially runtime performance).

One idea would be to nest the Enums:

```rs
enum SpotId {
    Amagi(AmagiSpotId),
    Antarctica(AntarcticaSpotId),
    ...
}
enum AmagiSpotId {
    Room1(AmagiRoom1SpotId),
    Room2(AmagiRoom2SpotId),
}
```
etc.

The benefit would be that our movement table would then just be within areas only, and some of the "upcast" functions like `get_area` only have to look at their variants. One downside is that this would be 3 or 4 bytes instead of just 2.

Plus, in AV2 this would turn from one enum with a few thousand entries to 307 area enums and 17 region enums and one top-level enum. If we only used the region level for indirection it would be a lot more manageable in terms of number of enums but that would have less benefit for the movement matrices.

We've already seen that replacing the EnumMap<SpotId, ContextWrapper> used in the access functions with a hashmap results in a 20-30% speedup for condensed maps, so we no longer have that concern.

**Comment by Zannick** [comment]
Functions/data that need to be replaced:
- [x] get_area
- [x] get_region
- [x] local_travel_time
- [ ] are_spots_connected
- [ ] base_edges
- [ ] free_movement
- [ ] best_movements
- [x] get_area_spots
- [ ] get_exit_spots
- [ ] get_all_spots
- [ ] spot_of_interest
- [ ] condensed graph type
- [x] build_spots
- [ ] get_spot_(points)
- [ ] action dest
- [ ] same_region?
- [ ] same_area?
- [ ] get_location_spot
- [ ] get_action_spot
- [ ] get_exit_spot
- [ ] context overrides

**Comment by Zannick** [comment]
I like some of what I've done here, and for anything not related to the new SpotId, I'll probably port it into main while we see if this is an improvement over that.

**Comment by Zannick** [comment]
So, it's a performance regression per `cargo bench`. I've tried to port over into main what I can that isn't part of the SpotId refactor, which may have resulted in some improvements on their own. My guess is that part of the issue is the inability to return `impl Iterator` from trait methods, which is why we are producing vectors before returning them to be processed. Last I checked, that feature is in nightly and added to the 1.75 milestone, so it could be worth revisiting when it reaches stable.

Additionally, the intent here was to reduce the memory usage of the compiler in debug without `overflow-checks = false`. Here, it's partially successful; I did observe compilation times decrease in the relevant areas, but we still spiked to 15 GiB. It's unclear if that's because we are still using `match (SpotId, SpotId)` statements (even though they only look at the discriminant), or because we added `get_local_spots` to simplify some of the local options.
